### PR TITLE
Fix issues with quotes and character encodings

### LINKED ISSUE #1: Problem with extended ASCII characters
First off: let me thank you for implementing this great app!

I've tried to decode some BAS files, which were created around 93-94 by my IT teacher Ágnes Drong in elementary school (Hungary, Budapest). She was also a math teacher that time, therefore she wrote mostly mathematics related educational programs.

I got some strange results, when I tried to decode them, as they contain special Hungarian characters from the extended ASCII set.

Example:

	$ ./convert.py matek/stul.bas
	   10 CLS
	   20 SCREEN 1:KEY OFF:H=0
	   30 LOCATE 12,8:PRINT"SokszCONTgek tulajdonsDIMgai"
	   40 X$=INPUT $(1):CLS
	   45 PRINT:PRINT:PRINT

There the text should be "Sokszögek tulajdonságai". (Properties of polygons).
I've attached all files: [matek.zip](https://github.com/danvk/gwbasic-decoder/files/3586485/matek.zip)

EDIT: In the meantime I've figured that these files used the encoding:
- Code page 437 (IBM PC)

### PR DESCRIPTION
**Pull request by bolner**
- Introduce 2 parsing states: inside quote, inside REM block
- Suppose 'IBM Code Page 437' encoding by default

Fixes #1

### CODE REVIEW

**Comment by danvk** [comment]
Hi @bolner, thanks for the fix! I'm glad you've gotten value out of this tool.

Would you mind:

1. Adding a test file with expected output that demonstrates the fix.
2. At least make the encoding a parameter an argument. You're right that it should be a flag, but then that means that there should be support for flags in general :)

**Comment by bolner** [comment]
Hi @danvk, yes, I'll do that today or tomorrow. Later I also realized other issues: like the single quote is also an alias for the REM. Plus: one of the files I attached is broken (probably data corruption due to age :) I'll try to fix it with a hex editor on the weekend), but the program doesn't detect it, it only displays the 20% of the lines (without error message). And I have a bad feeling about some of the floating point conversions. I think sometimes it would need rounding at less digits, to show the same numbers as the original were. I could test that in an emulator.
Also it doesn't run on Python 3, only on 2. Should that stay this way, or can I change it to Python 3? :) (I think it wouldn't support 2 anymore)

**Comment by danvk** [comment]
Feel free to make it Python 3. In fact, please make it Python 3!

**Comment by bolner** [comment]
@danvk Hi, I'm not ready yet, so I've created a dev branch: https://github.com/bolner/gwbasic-decoder/tree/dev

I've implemented the doubles, and fixed some float issues (The 0.8 in GAME.BAS is okay now), but rounding and formatting is still not okay everywhere. I've added a `TYPEDEMO.BAS` which demonstrates many types. Its output should be as on the screenshot I attached (It's from a real GW-Basic):
type_demo

So the hex / octal types are neither showing up properly.
I think I can finish it on the weekend :)

List of all changes:
- Changed the project to Python 3
- Committed a "launch.json" for VS Code, which can be used for debugging
- Added the encoding as a new command line parameter
- It outputs now usage info if called without parameters
- Added static typing
- Renamed the "bytes" variable to "data", so that variable won't have the same name as an existing type, which can be confusing.
- The data and the encoding are now passed to the objects in their constructors.
- Got rid of the static "factory methods".
- Rewrote the float conversion to avoid losing accuracy and implemented the double conversion.
- Renamed gwbasicline to gwBasicLine
- Moved the token parsing functions into gwBasicLine
- Renamed gwbasic to gwBasic
- It now can parse the single quote alias of the REM command (stored as 3 bytes in reality)
- The floats in GAME.BAS are now properly displayed. Updated the readme.

**Comment by bolner** [comment]
@danvk Now everything works which was planned, but I've noticed another bug. Did you know that the `GAME.BAS` has 7 more lines? :) Just checked it in an emulator. This problem existed before my changes as well. I'll check it.

**Comment by bolner** [comment]
@danvk Fixed everything, and tested with many BAS files, other than the ones in README.md. No more TODOs left :) Please have a look when you have some free time. Everything is in master.

File: convert.py [hunk 1]
```diff
@@ -1,13 +1,41 @@
...
 
 if __name__ == '__main__':
-    assert len(sys.argv) == 2
-    filename = sys.argv[1]
-    with open(filename, 'r') as f:
-        data = f.read()
-        program = gwbasic.FromBinary(data)
-        print program
+    # Check the number of parameters
+    if not (2 <= len(sys.argv) <= 3):
```
Comment (danvk) [inline convert.py]: Suggest making this a flag and using Python's built-in [argparse][1], which supports optional arguments, flags and values which have to be one of a set of choices. This sort of argument processing has a way of getting very messy as the number of parameters inevitably increases.

[1]: https://docs.python.org/3/library/argparse.html

File: convert.py [hunk 2]
```diff
@@ -1,13 +1,41 @@
...
+                raise ValueError("Unknown character encoding: %s" % sys.argv[2])
+        
+        filename = sys.argv[1]
+        with open(filename, 'rb') as f:
+            data = f.read()
+            program = gwBasic(data, encoding)
+            program.Parse()
+            print(program)
+
+    except Exception as ex:
```
Comment (danvk) [inline convert.py]: Suggest removing this try / except — it seems almost exactly the same as the behavior without it?

File: gwbasic.py [hunk 3]
```diff
@@ -1,144 +1,236 @@
...
+                # The boundary for that +1 is already checked at the beginning of the loop.
+                self.lineBuffer.append(tokens[(code << 8) | self.data[self.pos + 1]])
+                self.pos += 2
             else:
                 raise ValueError("unexpected token: %d" % code)
-        pos += 1    # consume the null byte
-        return [self, pos]
+
+        self.pos += 1    # consume the null byte
+        return
```
Comment (danvk) [inline gwbasic.py]: this is a no-op

File: gwbasic.py [hunk 4]
```diff
@@ -1,144 +1,236 @@
...
+            if code == 0x22 and not insideRem:  # Quote starts or ends
+                # There was no quote escaping. You had to use CHR$() to
+                #   output a quote character.
+                insideQuotes = not insideQuotes
+                self.lineBuffer.append('"')
+                self.pos += 1
+                continue
+            elif code == 0x3a and not (insideQuotes or insideRem):
+                if len(self.data) - self.pos - 1 > 2:
+                    if self.data[self.pos + 1] == 0x8f and self.data[self.pos + 2] == 0xd9:
```
Comment (danvk) [inline gwbasic.py]: combine these into a single `if`

File: gwbasic.py [hunk 5]
```diff
@@ -1,144 +1,236 @@
...
-    return (-1)**sgn * 2**exp * mantissa
-
-
-def ParseDouble(bytes):
-    return "<unparsed double>"
-    # return struct.unpack("!d", ''.join([chr(x) for x in bytes[8::-1]]))
-
-
-class gwbasicline:
+class gwBasicLine:
```
Comment (danvk) [inline gwbasic.py]: I realize you're inheriting this, but in Python you generally CamelCase class names, so `GWBasicLine` and `GWBasic`.

**Review by danvk — APPROVED** [review]
A few comments, but otherwise looks great. I'm going to add you as an owner of this repo since you've written more of the code than I have at this point!

File: gwbasic.py (hunk 3 above)
Comment (bolner) [reply gwbasic.py]: You mean the "return"? Yes, didn't notice, will remove it.

File: gwbasic.py (hunk 4 above)
Comment (bolner) [reply gwbasic.py]: Ah yes, I wasn't sure that Python supports this: https://en.wikipedia.org/wiki/Short-circuit_evaluation   Then we would also not need the 2 continues.

File: convert.py (hunk 2 above)
Comment (bolner) [reply convert.py]: Hmm yes :) Except for the extra newlines around it, but that might not worth the complication. Will remove.

File: convert.py (hunk 1 above)
Comment (bolner) [reply convert.py]: Just tried it, looks good, will commit soon.

**Comment by bolner** [comment]
@danvk Committed changes.

**Pull request merged**